### PR TITLE
Async/dynamic loader factories

### PR DESCRIPTION
**Pull request by ryantrem**
TLDR; the net effect of this change for the alpha Viewer is that it brings the **core bundle size down from ~1.5mb to ~1.15mb**. Anything else needed to load any type of supported file or glTF extension can be downloaded as a separate chunk on demand.

This is an alternative approach to #15492. Overall I think this one is much better. This is much more similar to what I originally wanted to do, but I was a bit stuck on the fact that the legacy loader functions synchronously returned the plugin instance. After talking with @sebavan though, we both agreed that since the user must explicitly opt into loaders anyway, it would be ok to have a new opt-in path for the dynamic loaders, and in this case, they give up the old functionality of having a plugin instance being synchronously returned from the legacy loader functions. So now with this change you can simply change:

```ts
import "@babylonjs/loaders/OBJ/objFileLoader";
```

to
```ts
import { registerOBJLoader } from "@babylonjs/loaders/OBJ/objFileLoader.dynamic";
registerOBJLoader();
```

Note that the new dynamic loaders could have worked via side effect just like the static ones, but we are trying to move away from side effects so I opted to require a function to be called.

When the dynamic loaders are used, you can still call the legacy loader functions, but they will return null instead of a plugin instance (the contract already allowed for the possibility of null being returned).

With this approach, `objFileLoader.dynamic` is 288 bytes minified, so very low overhead to be able to load the obj loader dynamically only when you actually try to load an obj file.

UPDATE: I've also now added the other loaders, and for glTF (2.0), extension factories can also now be async, which means they can also use dynamic imports. This also introduces the concept of a glTF loader extension either being associated with a glTF extension or not. If it is, then it is only instantiated if the extension is used. So in the case of dynamic imports, it will only be downloaded when needed. To maintain back compat, this requires splitting each existing extension file into at least two files - one that just has the implementation but not the side effect registration, and one for the side effect registration for back compat. For the one extension that I'm showing this pattern, I actually split it into three, where the third is the dynamic import async factory. If we don't want to split every extension into 3 files (rather just 2), then we could just mash all the factory registrations into a single dynamic.ts file. However, this would mean consumers could not choose to only include a subset of glTF extensions in their bundle chunks, so my personal preference is to just have 3 files. Future extensions could forego the old static registration pattern and only support dynamic registration, which would then mean 2 files per extension.

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15499/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15499/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15499/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15499/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15499/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15499/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15499/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15499/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15499/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15499/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15499/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15499/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by RaananW** [comment]
Loe this. this is a great idea for async-loading the needed loaders.
Will have to investigate what went wrong with the generation of the UMD declaration file.

File: packages/dev/loaders/src/SPLAT/splatFileLoader.plugin.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+     */
+    public readonly name = SPLATFileLoaderMetadata.Name;
+
+    /**
+     * Defines the extensions the splat loader is able to load.
+     * force data to come in as an ArrayBuffer
+     */
+    public readonly extensions = SPLATFileLoaderMetadata.Extensions;
+
+    //private _loadingOptions: SPLATLoadingOptions;
```
Comment (bghgary) [inline packages/dev/loaders/src/SPLAT/splatFileLoader.plugin.ts:34]: Is this supposed to be commented out?

Comment (ryantrem) [reply packages/dev/loaders/src/SPLAT/splatFileLoader.plugin.ts:34]: I didn't make any changes to this code, it's just moved from the .ts file to the .plugin.ts.

Comment (bghgary) [reply packages/dev/loaders/src/SPLAT/splatFileLoader.plugin.ts:34]: Yup, I see that. Maybe remove anyways since it's completely unreferenced.

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15499/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

File: packages/dev/loaders/src/SPLAT/splatFileLoader.plugin.ts (hunk 1 above)
Comment (ryantrem) [reply packages/dev/loaders/src/SPLAT/splatFileLoader.plugin.ts:34]: @CedricGuillemet will you fix this in your splat loader PR?
https://github.com/BabylonJS/Babylon.js/pull/15501

File: packages/tools/viewer-alpha/src/viewer.ts [hunk 2]
```diff
@@ -292,13 +296,13 @@ export class Viewer implements IDisposable {
 
         // TODO: Disable audio for now, later figure out how to re-introduce it through dynamic imports.
         options = {
-            ...options,
+            ...(options ?? {}),
```
Comment (sebavan) [inline packages/tools/viewer-alpha/src/viewer.ts:299]: the options management gets pretty hard to read.

File: packages/dev/loaders/src/SPLAT/splatFileLoader.plugin.ts (hunk 1 above)
Comment (CedricGuillemet) [reply packages/dev/loaders/src/SPLAT/splatFileLoader.plugin.ts:34]: yes!

File: packages/dev/loaders/src/glTF/2.0/Extensions/KHR_materials_pbrSpecularGlossiness.dynamic.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { GLTFLoaderExtensionRegistry } from "../glTFLoaderExtensionRegistry";
```
[the remark this answers is not in the document]
Comment (ryantrem) [reply packages/dev/loaders/src/glTF/2.0/Extensions/KHR_materials_pbrSpecularGlossiness.dynamic.ts:1]: That's what I was originally going to do, but because the extensions statically import the `GLTFLoader` which statically imports the `GLTFFileLoader`, it's basically not possible to have the `GLTFFileLoader`/`GLTFLoader` dynamically imported without dealing with the extensions. :/

File: packages/tools/viewer-alpha/src/viewer.ts (hunk 2 above)
Comment (ryantrem) [reply packages/tools/viewer-alpha/src/viewer.ts:299]: Yea merging options is pretty verbose. Since they are plain js objects, I'm not sure how to do this better, other than using an object merging library. That said, this specific code should go away once we support dynamically importing the audio engine. Also regardless of that, I'm pretty sure I will change this class more later to not even pass through `LoadAssetContainerOptions`. I think it will be better for this class to be more stateful and mimic the lifetime of the DOM layer to simplify the "binding" of the DOM layer (and any future other UI layers) to this Viewer layer.

File: packages/dev/loaders/src/dynamic.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * Registers the async plugin factories for all built-in loaders.
+ * Loaders will be dynamically imported on demand, only when a SceneLoader load operation needs each respective loader.
+ */
+export function registerBuiltInLoaders() {
+    // Register the glTF loader (2.0) specifically/only.
+    registerSceneLoaderPlugin({
+        ...GLTFFileLoaderMetadata,
+        createPlugin: async (options: SceneLoaderPluginOptions) => {
+            // eslint-disable-next-line @typescript-eslint/naming-convention
```
Comment (bghgary) [inline packages/dev/loaders/src/dynamic.ts:22]: remove?

File: packages/dev/loaders/src/glTF/2.0/Extensions/KHR_materials_pbrSpecularGlossiness.dynamic.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryantrem) [reply packages/dev/loaders/src/glTF/2.0/Extensions/KHR_materials_pbrSpecularGlossiness.dynamic.ts]: Talked offline, got aligned on extensions.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

File: packages/dev/loaders/src/glTF/2.0/glTFLoaderExtensionRegistry.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import type { GLTFLoader } from "./glTFLoader";
+import type { IGLTFLoaderExtension } from "./glTFLoaderExtension";
+
+import { Logger } from "core/Misc/logger";
+
+interface IRegisteredExtension {
```
Comment (bghgary) [inline packages/dev/loaders/src/glTF/2.0/glTFLoaderExtensionRegistry.ts:6]: nit:
```suggestion
interface IRegisteredGLTFExtension {
```

File: packages/dev/loaders/src/glTF/2.0/glTFLoaderExtensionRegistry.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import type { IGLTFLoaderExtension } from "./glTFLoaderExtension";
+
+import { Logger } from "core/Misc/logger";
+
+interface IRegisteredExtension {
+    isGLTFExtension: boolean;
+    factory: ExtensionFactory;
+}
+
+export type ExtensionFactory = (loader: GLTFLoader) => IGLTFLoaderExtension | Promise<IGLTFLoaderExtension>;
```
Comment (bghgary) [inline packages/dev/loaders/src/glTF/2.0/glTFLoaderExtensionRegistry.ts:11]: nit:
```suggestion
export type GLTFExtensionFactory = (loader: GLTFLoader) => IGLTFLoaderExtension | Promise<IGLTFLoaderExtension>;
```

File: packages/dev/loaders/src/glTF/2.0/glTFLoaderExtensionRegistry.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+const _glTFRegisteredExtensions = new Map<string, IRegisteredExtension>();
+
+/**
+ * All currently registered glTF 2.0 loader extensions.
+ */
+export const glTFRegisteredExtensions: ReadonlyMap<string, Readonly<IRegisteredExtension>> = _glTFRegisteredExtensions;
```
Comment (bghgary) [inline packages/dev/loaders/src/glTF/2.0/glTFLoaderExtensionRegistry.ts:18]: nit:
```suggestion
const _registeredGLTFExtensions = new Map<string, IRegisteredGLTFExtension>();

/**
 * All currently registered glTF 2.0 loader extensions.
 */
export const registeredGLTFExtensions: ReadonlyMap<string, Readonly<IRegisteredGLTFExtension>> = _registeredGLTFExtensions;

```

**Review by bghgary — APPROVED** [review]

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

File: package-lock.json [hunk 8]
```diff
@@ -24644,8 +24644,8 @@
                 "vite": "^5.4.2"
             },
             "peerDependencies": {
-                "@babylonjs/core": "^7.13.2",
-                "@babylonjs/loaders": "^7.0.0"
+                "@babylonjs/core": "^7.23.0",
```
Comment (sebavan) [inline package-lock.json:24647]: should it be the next version instead ? so 7.24.0 if merged tonight ?

**Review by sebavan — APPROVED** [review]

File: package-lock.json (hunk 8 above)
Comment (ryantrem) [reply package-lock.json:24647]: Yes, but I can't do that or it breaks the build. @RaananW says the only way is to put the wrong version, publish a build, fix the version, publish another build.

Comment (sebavan) [reply package-lock.json:24647]: oh yes cause it will install it before publishing at the end. so good luck with the double PR :-)

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15499/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.
